internal/observability.go

```diff
@@ -65,42 +67,111 @@ var ViewExportedSpans = &view.View{
+		predefinedSpanName: "opencensus.service.receiver." + receiverName + ".Receive",
+		startTime:          time.Now(),
+		receiverName:       receiverName,
+	}
+}
+
+// SpansObservabilityRecorder is a helper that is started when
+type SpansObservabilityRecorder interface {
+	Start(ctx context.Context, ni *commonpb.Node) context.Context
+	End(ctx context.Context, nItems int64, err error) context.Context
+	UnderlyingSpan() *trace.Span
+}
+
+type spansRecorder struct {
+	predefinedSpanName string
+	span               *trace.Span
+	receiverName       string
+	exporterName       string
+	startTime          time.Time
 }
 
-// ContextWithReceiverName adds the tag "opencensus_receiver" and the name of the
-// receiver as the value, and returns the newly created context.
-func ContextWithReceiverName(ctx context.Context, receiverName string) context.Context {
-	ctx, _ = tag.New(ctx, tag.Upsert(tagKeyReceiverName, receiverName))
+var _ SpansObservabilityRecorder = (*spansRecorder)(nil)
+
+func (sr *spansRecorder) Start(ctx context.Context, ni *commonpb.Node) context.Context {
+	sr.startTime = time.Now()
+	ctx, sr.span = trace.StartSpan(ctx, sr.predefinedSpanName, trace.WithSampler(trace.NeverSample()))
+
+	if sr.receiverName != "" {
+		ctx, _ = tag.New(ctx, tag.Upsert(tagKeyReceiverName, sr.receiverName))
+	} else {
+		ctx, _ = tag.New(ctx, tag.Upsert(tagKeyExporterName, sr.exporterName))
+	}
+
 	return ctx
 }
 
-// NewReceivedSpansRecorderStreaming creates a function that uses a context created
-// from the name of the receiver to record the number of the spans received
-// by the receiver.
-func NewReceivedSpansRecorderStreaming(lifetimeCtx context.Context, receiverName string) func(*commonpb.Node, []*tracepb.Span) {
-	// We create and reuse this context because for streaming RPCs e.g. with gRPC
-	// the context doesn't change, so it is more useful for avoid expensively adding
-	// keys on each invocation. We can create the context once and then reuse it
-	// when recording measurements.
-	ctx := ContextWithReceiverName(lifetimeCtx, receiverName)
-
-	return func(ni *commonpb.Node, spans []*tracepb.Span) {
-		// TODO: (@odeke-em) perhaps also record information from the node?
-		stats.Record(ctx, mReceivedSpans.M(int64(len(spans))))
+func (sr *spansRecorder) End(ctx context.Context, nItems int64, err error) context.Context {
+	var measurements []stats.Measurement
+	var tagMutators []tag.Mutator
+
+	if sr.receiverName != "" {
+		tagMutators = append(tagMutators, tag.Upsert(tagKeyReceiverName, sr.receiverName))
+		measurements = append(measurements, mReceivedSpans.M(nItems))
+	} else {
+		tagMutators = append(tagMutators, tag.Upsert(tagKeyExporterName, sr.exporterName))
+		measurements = append(measurements, mExportedSpans.M(nItems))
+	}
+
+	if err == nil {
+		tagMutators = append(tagMutators, tag.Upsert(tagKeyStatus, "OK"))
+	} else {
+		tagMutators = append(tagMutators, tag.Upsert(tagKeyStatus, "ERROR"), tag.Upsert(tagKeyError, err.Error()))
```

Comment:
error is an int without any bounds, I don't think it is fine to put it as a tag in this case, that is the main reason we defined a type of errors (see the status in trace).

I would suggest to use if possible the same mapping that we have in trace (see https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto)